### PR TITLE
feat(client): Log warning in resend subscribe if no storage assigned

### PR DESCRIPTION
**Pull request by teogeb**
If storage node is not assigned to a stream and we call `client.resendSubscribe`, it logs a warning. Discussed here: https://github.com/streamr-dev/network-monorepo/pull/921#issuecomment-1288784070

## Testing

Added `FakeLogger` test environment class. It stores log entries to an array, and in test we can check that an expected entry exists in that array.

Created a type annotation for `LogLevel` and used that in `FakeLogger` and `StrictStreamrClientConfig`.

### CODE REVIEW

File: packages/client/test/unit/ResendSubscription.test.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import { ResendSubscription } from "../../src/subscribe/ResendSubscription"
+import { fromArray } from '../../src/utils/GeneratorUtils'
+import { collect } from '../../src/utils/iterators'
+import { STREAM_CLIENT_DEFAULTS } from './../../src/Config'
+import { StreamrClientError } from './../../src/StreamrClientError'
+import { createMockMessage } from './../test-utils/utils'
+
+const STREAM_PART_ID = StreamPartIDUtils.parse('streamId#0')
+
+describe('ResendSubscription', () => {
```
Comment (harbu) [inline packages/client/test/unit/ResendSubscription.test.ts]: Check logging in Resends2.test.ts with integration test as discussed.

Comment (teogeb) [reply packages/client/test/unit/ResendSubscription.test.ts]: Fixed

File: packages/client/test/test-utils/fake/FakeLogger.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    message: string
+    level: LogLevel
+}
+
+const LOG_LINE_PREFIX_LENGTH = 56 // Logger prefixes each line with level, timestamp and context
+
+export class FakeLogger implements Omit<Logger, 'getFinalLogger'> {
+
+    private entries: LogEntry[] = []
+    private format: (pattern: string, args: any[], cb: (message: string) => void) => void
```
Comment (harbu) [inline packages/client/test/test-utils/fake/FakeLogger.ts]:
```suggestion
    private readonly entries: LogEntry[] = []
    private readonly format: (pattern: string, args: any[], cb: (message: string) => void) => void
```

File: packages/client/test/test-utils/fake/FakeLogger.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        this.format = (pattern: string, args: any[], cb: (message: string) => void) => {
+            const logger = new Logger(module, undefined, 'trace', {
+                write: (line: string) => {
+                    // eslint-disable-next-line
+                    const withoutColors = line.replace(/\x1b\[[0-9]+m/gi, '')
+                    const withoutPrefix = withoutColors.substring(LOG_LINE_PREFIX_LENGTH)
+                    const withoutLinefeed = withoutPrefix.substring(0, withoutPrefix.length - 1)
+                    cb(withoutLinefeed)
+                }
+            })
```
Comment (harbu) [inline packages/client/test/test-utils/fake/FakeLogger.ts:26]: I think we should create a simple unit test for `FakeLogger` to ensure that it strips away this info correctly. To guard against situations where we would change the format of the real `Logger` but forget to account for the change here.

File: packages/client/test/test-utils/fake/FakeEnvironment.ts [hunk 4]
```diff
@@ -79,6 +87,10 @@ export class FakeEnvironment {
         return this.network
     }
 
+    getLogger(): FakeLogger {
+        return this.logger
+    }
+
```
Comment (harbu) [inline packages/client/test/test-utils/fake/FakeEnvironment.ts:93]: Should we provide a direct access to log entries here to make use easier?
```suggestion
    getLogEntries(): LogEntry[] {
        return this.logger.getEntries()
    }

```

File: packages/client/test/test-utils/fake/FakeLogger.ts (hunk 2 above)
Comment (teogeb) [reply packages/client/test/test-utils/fake/FakeLogger.ts]: Fixed

File: packages/client/test/test-utils/fake/FakeLogger.ts (hunk 3 above)
Comment (teogeb) [reply packages/client/test/test-utils/fake/FakeLogger.ts:26]: Added a unit test

File: packages/client/test/test-utils/fake/FakeEnvironment.ts (hunk 4 above)
Comment (teogeb) [reply packages/client/test/test-utils/fake/FakeEnvironment.ts:93]: I'd keep the current style where each `FakeEnvironment` component is exposed as a getter, and the actual query/command methods are in that component. It is quite straightforward to assert `environment.getLogger().getEntries()` anyway.

This way we follow the "do only one thing, and do it well" principle in `FakeEnvironment`: the environment just manages the component, doesn't operate on them (and doesn't need to know what operations are available).

**Review by harbu — APPROVED** [review]

**Pull request merged**